storage: Added support for chunk iterators: Remote read client, readyStorage, fanout.

bwplotka opened the pull request:
## Major Changes

* Added ChunkQueryable implementation to fanout and readyStorage.
* Added more comments.
* Changed NewVerticalChunkSeriesMerger to CompactingChunkSeriesMerger, removed tiny interface by reusing VerticalSeriesMergeFunc for overlapping algorithm for
both chunks and series, for both querying and compacting (!) + made sure duplicates are merged.
* Added ErrChunkSeriesSet
* Added Samples interface for seamless []promb.Sample to []tsdbutil.Sample conversion.
* Deprecating non chunks serieset based StreamChunkedReadResponses, added chunk one.
* Improved tests.
* Split remote client into Write (old storage) and read.
* Queryable client is now SampleAndChunkQueryable. Since we cannot use nice QueryableFunc I moved
all config based options to sampleAndChunkQueryableClient to aboid boilerplate.

In next commit: Changes for TSDB.

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

brian-brazil commented on tsdb/db.go:
```diff
@@ -1311,17 +1310,45 @@ func (db *DB) Querier(_ context.Context, mint, maxt int64) (storage.Querier, err
+}
+
+// Querier returns a new querier over the data partition for the given time range.
+// A goroutine must not handle more than one open Querier.
```
I don't think this is what you meant to say

bwplotka replied:
Copied comment, sorry. FIxing.

bwplotka replied:
If you mean

 > // A goroutine must not handle more than one open Querier.

That was 1:1 copied from db.go ;p  Since I don't get what it means, I will remove it.

brian-brazil commented on storage/fanout.go:
```diff
@@ -696,32 +720,29 @@ func (h *samplesIteratorHeap) Pop() interface{} {
+//
+// TODO(bwplotka): Currently we have only ChainingSeriesMerge as overlapSeriesMergeFunc. This works the best with
+// replicated series, where data from two series are exactly the same. This does not work well with "almost" the same
+// data e.g from 2 Prometheus HA replicas. Propose penalty based algorithm from Thanos as alternative.
```
From a Prometheus standpoint, non-identical data is not a sane setup for remote read. A merge must be a merge, without losing samples.

bwplotka replied:
Yes agree. I just propose to add some alternative that will work for all cases. Will add your mentioned use case to comment.

brian-brazil replied:
We shouldn't be documenting Thanos design decisions that don't apply to Prometheus inside Prometheus.

bwplotka replied:
Generally agree, but what if someone want to merge two blocks from different Prometheus replicas (fair use case) without even Thanos?

bwplotka replied:
I can remove Thanos mention but will mention about some potentiall improvement we can make for such use case, WDYT?

brian-brazil replied:
Then they're going to have unsuable data, that it may be inefficient doesn't really matter as the use case isn't sane.

bwplotka replied:
Fair. Added comment to `ChainingSeriesMerge` instead. (:

bwplotka commented on storage/fanout.go:
```diff
@@ -91,7 +91,33 @@ func (f *fanout) Querier(ctx context.Context, mint, maxt int64) (Querier, error)
+	return NewMergeQuerier(primaryQuerier, queriers, ChainingSeriesMerge), nil
+}
+
+func (f *fanout) ChunkQuerier(ctx context.Context, mint, maxt int64) (ChunkQuerier, error) {
```
This deserves a separate file, but will move in next PR to have some diff here.

bwplotka commented:
Ok as proposed by @codesome offline I split this into one piece here to fix and prepare storage and remote packages. This is follow up to implement ChunkQuerier in tsdb and unify series set merges: https://github.com/prometheus/prometheus/pull/7069.

Will fix tests for both PRs  tomorrow

codesome commented:
Thanks for splitting!

bwplotka commented:
All tests should be fixed and this PR should be ready for careful review and merge :hugs: 

cc @brian-brazil @codesome @brian-brazil

bwplotka commented on storage/remote/read_test.go:
```diff
@@ -231,182 +211,310 @@ func TestSeriesSetFilter(t *testing.T) {
 			matchers: []*labels.Matcher{
-				labels.MustNewMatcher(labels.MatchRegexp, "special", "label"),
+				labels.MustNewMatcher(labels.MatchNotEqual, "a", "something"),
+				// TODO(bwplotka): Is this what we want? I prometheus is region=europe and user requests region=us we
```
A bit surprised with this. Is this intended (this PR does not change any semantic, just I added a test case for this), @brian-brazil ? :thinking:

bwplotka commented:
Also cc @pstibrany and @pracucci

brian-brazil commented on storage/remote/client.go:
```diff
@@ -38,30 +39,71 @@ const maxErrMsgLen = 256
+		},
+	)
+	if reg != nil {
+		reg.Unregister(readQueries)
```
This doesn't seem right, you should never have to unreigster outside of tests.

bwplotka replied:
Hm, what're rationales behind this?

Because otherwise I would need to move this gauge to storage (which is fair I guess)

brian-brazil replied:
A given metric should exist for the lifetime of a process, if you're unregistering then that implies something is off with the metric lifecycle.

bwplotka replied:
oke will fix

brian-brazil commented on storage/fanout.go:
```diff
@@ -190,66 +215,80 @@ func (f *fanoutAppender) Rollback() (err error) {
+			continue
+		}
+		if i == primaryQuerier {
+			// Primary was filtered, no primary anymore.
```
How can this happen?

bwplotka replied:
Well any consumer of this function can do this anytime, but just specifying `MergeQuerier([]Querier{NoopQuerier()}, 0, ...)` no?

brian-brazil replied:
Do we have any of those in Prometheus, or are you keeping this general for Thanos?

bwplotka replied:
We don't have at the current moment, but what if write code that does it? We don't use this at all for Thanos.

brian-brazil replied:
We can worry about it then, as-is we always have a primary querier and it's always present.

brian-brazil replied:
The primaryQuerier argument to mergeGenericQuerier seemed unneeded now, and we're not handling it correctly now anyway.

brian-brazil replied:
This still doesn't look correct.

bwplotka replied:
looking

bwplotka replied:
Hm, what do you mean?

PrimaryQuerier argument actually tells us if it should be error or warning e.g:

```
	           if q.primaryQuerier == 0 {
				return nil, w, err
			}
			// Nil will be converted to noop in adapter.
			return nil, append(w, err), nil
```

and

```
                if r.selectError != nil {
			q.failedQueriers[q.queriers[r.i]] = struct{}{}
			if r.i != q.primaryQuerier {
				// If the error source isn't the primary querier, return the error as a warning and continue.
				warnings = append(warnings, r.selectError)
				continue
			}
			priErr = r.selectError
			continue
		}
```
So it's used, and to me it looks correctly.

bwplotka replied:
@brian-brazil

bwplotka replied:
What's broken it's that the first if `len(queries) == 1` is not needed.

bwplotka replied:
Removed.

brian-brazil replied:
This still doesn't look right (warnigns vs errors), and even if it is it's going to be fragile. I think passing the primary querier as an index rather than the actual object is the fundamental issue, so would suggest going back to having a primary querier and a list of secondaries.

bwplotka replied:
Hm tend to agree with this, but the previous logic was extremely surprising as we expected caller to fill all queries + primary additionally.

I think I would be fine with primary and secondaries as different args, I think that would  work for both of all (:

bwplotka commented:
Ok, fixed @brian-brazil  (:

bwplotka commented:
Friendly ping @brian-brazil @krasi-georgiev (:

Sudhar287 commented:
Hello bwplotka!
You had [asked](https://github.com/prometheus/prometheus/pull/6760#pullrequestreview-383385739) me to work on the `Iterator` function in `storage/series.go`. 
I think a lot of my code from that PR can simply be moved over to this one. 
I have a question though. Are you planning to commit the basic structure of the test code (in `storage/series_test.go`) you would like me to follow for the `Iterator` function? I'm asking this because I think it's better to have the structure that you expect rather than going back and forth in a couple of PRs. :) Any examples from other Prometheus test files that I should follow?

bwplotka commented:
@Sudhar287 I `proposed` rather. (: 

But I think I have a better idea. Looks regarding this code it's a bit risky as it's still in review and can change. If you want to help I would recommend more important feature: https://github.com/prometheus/prometheus/issues/5926  WDYT? (: 

Friendly ping for review @brian-brazil @codesome @krasi-georgiev, this PR sits here for a week already.

brancz commented on storage/fanout.go:
```diff
@@ -190,48 +215,54 @@ func (f *fanoutAppender) Rollback() (err error) {
+// NewMergeQuerier returns a new Querier that merges results of given slice of queriers.
+// NewMergeQuerier will return NoopQuerier if no queriers or noopQueriers are passed.
+// Function allows to optionally mark certain querier in the slice to be primary one. The difference between primary and
+// rest is as follows: if the primary querier returns an error, query fails. For secondaries it just return warnings.
```
"query fails" -> "the query fails"

"it just return" -> "it just returns"

brancz commented on storage/fanout.go:
```diff
@@ -190,48 +215,54 @@ func (f *fanoutAppender) Rollback() (err error) {
+// Argument 'primaryQuerier' set to -1 means no primary querier.
+
+// In case of overlaps in the returned data by all queriers, mergeFunc will be used.
+func NewMergeQuerier(queriers []Querier, primaryQuerier int, mergeFunc VerticalSeriesMergeFunc) Querier {
```
this seems like quite an odd API, why not just pass the primary querier separately from the secondary ones?

bwplotka replied:
We discussed it here: https://github.com/prometheus/prometheus/pull/7059#discussion_r400914446

I think at the end that's the plan, will update in ~2h/tomorrow (:

bwplotka replied:
Wonder if primary should be slice for consistency... Maybe for now as we don't need that, let's stick to one.

brian-brazil replied:
Prometheus only has one local tsdb.

brancz commented on storage/remote/client.go:
```diff
@@ -38,30 +39,54 @@ const maxErrMsgLen = 256
 
-// NewClient creates a new Client.
-func NewClient(remoteName string, conf *ClientConfig) (*Client, error) {
+// TODO(bwplotka): Add streamed chunked remote read (add issue link).
```
did you already open this issue?

bwplotka replied:
yes, will update link: https://github.com/prometheus/prometheus/issues/5926 forgot

brancz commented on storage/remote/read.go:
```diff
@@ -219,6 +199,40 @@ func (q externalLabelsQuerier) addExternalLabels(ms []*labels.Matcher) ([]*label
 	return ms, el
 }
 
+// LabelValues implements storage.Querier and is a noop.
```
I can't find it right now but I feel like I've seen an issue or WIP PR about this.

bwplotka replied:
ack will search and link

brancz commented on storage/series.go:
```diff
@@ -128,11 +138,83 @@ func (c *chunkSetToSeriesSet) Err() error {
+}
+
+// TODO(bwplotka): Currently encoder will just naively build one chunk, without limit. Split it: https://github.com/prometheus/tsdb/issues/670
+// DO IT NOW.
```
:thinking: ?

bwplotka replied:
XD talking to myself :man_facepalming: 

will remove.

brancz commented:
Largely looks reasonable. Have we run prombench on this?

bwplotka commented:
>  Have we run prombench on this?

Nope, let's run on third PR to include whole feature maybe?

brancz commented:
That works for me.

bwplotka commented:
All should be addressed here, please also look on the next one: https://github.com/prometheus/prometheus/pull/7069

brian-brazil commented on storage/remote/client.go:
```diff
@@ -38,30 +39,54 @@ const maxErrMsgLen = 256
-// NewClient creates a new Client.
-func NewClient(remoteName string, conf *ClientConfig) (*Client, error) {
+// TODO(bwplotka): Add streamed chunked remote read method as well (https://github.com/prometheus/prometheus/issues/5926).
+type ReadClient interface {
```
Public interfaces should have comments.

brian-brazil commented on storage/remote/client.go:
```diff
@@ -38,30 +39,54 @@ const maxErrMsgLen = 256
+
+// NewReadClient creates a new client for remote read.
+// Caller has to ensure there are no duplicated clients.
+func NewReadClient(readQueries prometheus.Gauge, name string, conf *ClientConfig) (ReadClient, error) {
```
Should we pass in a registry instead?

bwplotka replied:
lemme see

bwplotka replied:
Ok, so no, in this case metric is what we want because we can create and delete clients for the whole life of the process and as you commented that we cannot use `Unregister` it's more explicit and nicer to register those on Storage level which is singleton.

brian-brazil replied:
It still feels weird to be passing metrics around. What happens if we add more metrics such as latency around?

bwplotka replied:
Why? We use that pattern in many many places in different projects to allow sharing the same `GaugeVec`. If we will have more than one we can create `readClientStruct` like we do in many places in Prometheus as well (:

brian-brazil replied:
A given metric should only usually exist in a single function and file, if you're passing it around that indicates that the factoring probably isn't correct.

bwplotka replied:
This is not usual then. (: It's part of bigger system. :thinking: Let's talk offline, I really feel strong opinion that this is better than global metric. Cannot really see alternative here.

@beorn7 any feedback on this?

brian-brazil replied:
Yes, and its lack of usualness is my concern. Instantiators of a class shouldn't care about what metrics it defines.

bwplotka replied:
That is fair statement. OK, I think I know the solution, let's just not export this. It is not used anywhere, just in `Storage` type. We can also have no constructor as it's constructed just once. I believe this will not violate `Instantiators of a class shouldn't care about what metrics it defines.` as it will internal type, tied to storage, in the same some `seriesSet` might have tied `seriesSetIterator` in it. WDYT?

bwplotka replied:
Changed, so it is not exported anymore, thus tied to stoage.

brian-brazil replied:
Still seems a bit weird, visibility doesn't really change things.

beorn7 replied:
I think you can pass around a registry and then either register a `Gauge` with the `remoteName` and `endpoint` labels as const labels (the classical approach) or use [WrapRegistererWith](https://pkg.go.dev/github.com/prometheus/client_golang@v1.5.0/prometheus?tab=doc#WrapRegistererWith) for the same effect.

bwplotka replied:
Well that's the problem. None of those solutions works because we can create many clients of same labels for lifetime of the process (configure remote read for server example.com, then remove then configure again).

bwplotka replied:
That would help indeed. But my question is.. why complicating things? We use the same pattern as I suggest in a dozen of places in Thanos and it works really well. What's wrong with it? (: 

Anyway, don't mind to change to AreadyRegisteredError just curious.

brian-brazil replied:
If a you get a AreadyRegisteredError I'd consider that a programming error and if you get it on purpose, poor architecture as you're trying to re-use metrics rather than define the right one once.

bwplotka replied:
Done. Also talked offline https://matrix.to/#/!WaUKIfoqfiyWQhenET:matrix.org/$15869391701329748dlPKk:matrix.org?via=matrix.org

Looks like it looks not bad,  happy with that. Thanks for discussion @brian-brazil and @beorn7. PTAL (:

brian-brazil replied:
As indicated above, I consider any use of AreadyRegisteredError to be incorrect.

bwplotka replied:
I have to reuse one... that's the whole point.

bwplotka replied:
Ok at this point ... we are blocked. So far @brian-brazil  you did not propose ANY working idea and all my 4 attempts (Unregister, passing gauge, passing gauge on a private struct, AlreadyExistsError different) you rejected. So... what we do know? Because I am 100% sure the version with global is wrong. 

What would you suggest then?

brian-brazil replied:
They way I'd approach this is to have a statically defined metric, pass down just the label value down from the instantiator, and have the instance access the child it needed via that label value.

brian-brazil replied:
Search for Publisher on https://www.robustperception.io/label-lookups-and-the-child for the basic pattern I'd use.

bwplotka replied:
Hah that's interesting as the child pattern is exactly what I did with `newReadClient(readQueries.WithLabels(..., ...), name, ...)`, just the parent metric was not global/static (package level), but from parent struct. (:

bwplotka replied:
Otherwise all of it is package level so not really good solution.

bwplotka replied:
Anyway, looks like there is no agreement here, so, lets do this, let's choose whatever now from the 5 attempts we had and I will start separate issue with some options how we can fix many package level metric definitions (and If we want to fix it) in one go (:  WDYT? I can even revert to package level even though I think it's a bad solution.

beorn7 replied:
I find the (package-)global variable the most problematic of all suggested ways. It's not only a style question, it just hard-breaks if you import the package multiple times with different import paths (as it is common in Go anyway, but now also a core concept of Go modules).

I see drawbacks of both the AlreadyExistsError pattern and the unregister approach, but both are not too bad IMHO.

bwplotka replied:
Thanks. I am opening a separate issue about it will summarize our findings.

Let's focus on iterating over chunks here :hugs:

brian-brazil replied:
> as the child pattern is exactly what I did with

No, that's quite different as you're passing the metric rather than the label value. Metric objects should not pass API boundaries.

> I see drawbacks of both the AlreadyExistsError pattern and the unregister approach

I think that any approach that relies on known anti-patterns is not a good approach, and we should not be encouraging such things by putting them inside Prometheus itself. Users looking to use these are usually either instrumenting at the wrong level, or should be writing a custom collector instead.

> I can even revert to package level even though I think it's a bad solution.

It's what the code already did, and it's a fine pattern that won't cause us any problems.

beorn7 replied:
> It's what the code already did, and it's a fine pattern that won't cause us any problems.

It's an anti-pattern, and it's fundamentally broken in any non-trivial code.

brian-brazil commented on storage/fanout.go:
```diff
@@ -69,29 +69,47 @@ func (f *fanout) StartTime() (int64, error) {
-			for _, q := range queriers {
+			for _, q := range secondaries {
+				// TODO(bwplotka): Log error.
+				_ = q.Close()
```
Don't you need to close the primary too?

bwplotka replied:
Nice catch!

brian-brazil commented on storage/fanout.go:
```diff
@@ -69,29 +69,47 @@ func (f *fanout) StartTime() (int64, error) {
+		if err != nil {
+			for _, q := range secondaries {
 				// TODO(bwplotka): Log error.
 				_ = q.Close()
```
Ditto

brian-brazil commented on storage/fanout.go:
```diff
@@ -188,142 +206,146 @@ func (f *fanoutAppender) Rollback() (err error) {
-		return q.queriers[0].Select(sortSeries, hints, matchers...)
-	}
-
+// NOTE: mergeGenericQuerier selects always return series sorted no matter if sorting was not required from caller.
```
I can see people depending on this when they shouldn't as it's a function comment, I'd put this inside the function as an implementation note.

bwplotka replied:
ack

bwplotka commented:
React issue on CI

brian-brazil commented on storage/fanout.go:
```diff
@@ -323,6 +324,8 @@ func (q *mergeGenericQuerier) Select(_ bool, hints *SelectHints, matchers ...*la
 		return nil, warnings, priErr
 	}
+
+	// NOTE: newGenericMergeSeriesSet always return series sorted.
```
No need for NOTE, it's not that important a detail.

brian-brazil commented on storage/remote/client.go:
```diff
@@ -56,6 +56,7 @@ type ClientConfig struct {
 	HTTPClientConfig config_util.HTTPClientConfig
 }
 
+// ReadClient allows to use SAMPLES method read against remote server.
```
uses the SAMPLES

It doesn't permit it, it does it.

bwplotka replied:
;po

brian-brazil commented on storage/remote/read_test.go:
```diff
@@ -350,7 +350,7 @@ func TestSampleAndChunkQueryableClient(t *testing.T) {
 			matchers: []*labels.Matcher{
 				labels.MustNewMatcher(labels.MatchNotEqual, "a", "something"),
-				// TODO(bwplotka): Is this what we want? I prometheus is region=europe and user requests region=us we
+				// TODO(bwplotka): Is this what we want? If prometheus is in region=europe and user requests region=us we
```
Yes, this is the desired behaviour. US&Europe isn't very sane as they're at the same level, but think about a hierarchical setup.

bwplotka replied:
ok

bwplotka commented:
@brian-brazil addressed (also rebased - to hope React unit test was fixed)

bwplotka commented:
Done @brian-brazil (:

bwplotka commented:
I think this is slowly ready to be merged, but happy to fix https://github.com/prometheus/prometheus/pull/7069 (fixing last nit) chained PR and run prombench on both. :hugs:

codesome commented on storage/fanout.go:
```diff
@@ -188,142 +208,146 @@ func (f *fanoutAppender) Rollback() (err error) {
+		close(resultChan)
+	}()
+
+	for r := range resultChan {
```
Few questions here:
1. Don't we need this to be in time-order as we use this with chained merge too?
2. (unrelated to this line) Do we attempt vertical merge for the queries which do not have overlapping data too?

bwplotka replied:
1. We don't.  `ChainingSeriesMerge` handles this just fine. But good point to add note to `mergeFunc` that it has to handle underodered time wise series (: Thanks! 
2. No, see:

```
func (c *genericMergeSeriesSet) At() Labels {
	if len(c.currentSets) == 1 {
		return c.currentSets[0].At()
	}
	series := make([]Labels, 0, len(c.currentSets))
	for _, seriesSet := range c.currentSets {
		series = append(series, seriesSet.At())
	}
	return c.mergeFunc(series...)
}

```

codesome commented on storage/fanout.go:
```diff
@@ -569,11 +593,14 @@ func (h *genericSeriesSetHeap) Pop() interface{} {
 
-// ChainedSeriesMerge returns single series from many same series by chaining samples together.
+// ChainingSeriesMerge returns single series from many same series by chaining samples together.
 // In case of the timestamp overlap, the first overlapped sample is kept and the rest samples with the same timestamps
```
> the first overlapped sample is kept

I just noticed this. In that case, we need to be consistent (maybe a future PR for someone). When vertical queries and compaction was introduced, the design was to keep the last appearing sample.

bwplotka replied:
Will check thanks!

bwplotka replied:
I mean... what is last and first? It's really random and always was... which one is first which last ;p  Maybe not from algo perspective but from the caller perspective - we randomly put results from different remote APIs (in random order)

bwplotka replied:
Changed docs to have "random one"

codesome replied:
If we sort the blocks based on their minTime, then the block with lower mint is first and higher is the last - that was what was used in compaction and vertical query. Because we use heap, I don't think that is easy to take care of it.

codesome commented on storage/fanout.go:
```diff
@@ -569,11 +593,14 @@ func (h *genericSeriesSetHeap) Pop() interface{} {
+// This works the best with replicated series, where data from two series are exactly the same. This does not work well
+// with "almost" the same data, e.g. from 2 Prometheus HA replicas. This is fine, since from the Prometheus perspective
+// this never happens.
+func ChainingSeriesMerge(s ...Series) Series {
```
While the name was changed back in #7005, I see that this also merges the overlapping data. Chaining sounds more like non-overlapping and attaching results directly.

bwplotka replied:
Cool, so what would you suggest? Because to me this algo is exactly doing that. It does not merge anything.. It's just appends ALL together not caring about any scrape interval at all....

codesome replied:
https://github.com/prometheus/prometheus/blob/a2000f39a70983ec1f25fa0c6bc43e68a6384d1f/storage/fanout.go#L681-L685

This tells me that we are actually vertically merging it here by dropping the samples with same timestamp.

codesome commented:
(My chrome was choking with the large diff open :) )

bwplotka commented on storage/fanout.go:
```diff
@@ -188,142 +208,146 @@ func (f *fanoutAppender) Rollback() (err error) {
-		return primaryQuerier
-	}
-
-	if primaryQuerier == nil && len(filtered) == 1 {
```
See: https://github.com/prometheus/prometheus/pull/7059#discussion_r400914446

bwplotka replied:
This is actually good point, I am adding about improvements to https://github.com/prometheus/prometheus/pull/7251 let's get this quick path back.

bwplotka commented on storage/fanout.go:
```diff
@@ -69,29 +69,49 @@ func (f *fanout) StartTime() (int64, error) {
-			for _, q := range queriers {
+			_ = primary.Close()
+			for _, q := range secondaries {
+				// TODO(bwplotka): Log error.
```
Let's decide on next PR, no regression here.

bwplotka commented:
Addressed all @codesome :heart: 

Rebased and fixed all. Continuing tomorrow on finishing part 3 finally to make sure we can run prombench on all: https://github.com/prometheus/prometheus/pull/7069

brian-brazil commented:
:+1:

codesome commented on tsdb/db.go:
```diff
@@ -39,6 +39,7 @@ import (
 	"github.com/prometheus/prometheus/tsdb/chunkenc"
 	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
 	"github.com/prometheus/prometheus/tsdb/fileutil"
+
```
```suggestion
```

bwplotka replied:
`goimports does this

bwplotka commented:
@codesome I renamed `ChainingSeriesMerge` to `OverlappedSeriesMerge` I think you point was good, we should make it clear

kakkoyun commented on storage/fanout.go:
```diff
@@ -188,142 +208,145 @@ func (f *fanoutAppender) Rollback() (err error) {
+			// We need to sort for newGenericMergeSeriesSet to work.
+			set, wrn, err := q.Select(true, hints, matchers...)
+			resultChan <- &queryResult{q: q, set: set, wrn: wrn, err: err}
+			wg.Done()
```
In case the `q.Select` implementation panics, maybe it should be a deferred call. `defer wg.Done()`

bwplotka replied:
if it panics then all panics anyway, but defer sounds good

bwplotka commented:
:wave: Rebased and updated all - now will do part 2 in next PR so we can run prombench :hugs: 

Also fixed small nits introduced by https://github.com/prometheus/prometheus/pull/7334

bwplotka commented on storage/remote/client.go:
```diff
@@ -42,37 +42,86 @@ const maxErrMsgLen = 256
-// Client allows reading and writing from/to a remote HTTP endpoint.
-type Client struct {
+func init() {
+	prometheus.MustRegister(remoteReadQueriesTotal, remoteReadQueries, remoteReadQueryDuration)
```
I just move the code don't laugh! ;p

bwplotka commented on storage/remote/client.go:
```diff
@@ -42,37 +42,86 @@ const maxErrMsgLen = 256
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Subsystem: subsystem,
+			Name:      "remote_read_queries",
```
With what? I think I would try to not change things from outside, that would mean some compatibility issue potentially (:

brian-brazil commented on web/api/v1/api.go:
```diff
@@ -1215,78 +1216,142 @@ func (api *API) remoteRead(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/x-protobuf")
+	w.Header().Set("Content-Encoding", "snappy")
+
+	// On empty or unknown types in req.AcceptedResponseTypes we default to non streamed, raw samples response.
```
This comment would make more sense up where this function is called.

bwplotka replied:
Will fix in next PR :hugs:

bwplotka replied:
This one: https://github.com/prometheus/prometheus/pull/7069
